### PR TITLE
Dynamicinput

### PR DESCRIPTION
**Pull request by nicolas-chaulet**
What's new:
- Simpler model building logic
- Model takes a whole dataset as input
- Moved methods from dataset specific implementations to the BaseDataset class
- Pointnet and KPConv can take features as an input as well as just points

### CODE REVIEW

File: datasets/base_dataset.py [hunk 1]
```diff
@@ -30,31 +30,37 @@ def __init__(self, dataset_opt, training_opt):
...
-    def input_dim(self):
-        pass
+    def feature_dimension(self):
+        return self._feature_dimension
+
+    @staticmethod
+    def extract_point_dimension(dataset: Dataset):
+        sample = dataset[0]
+        if sample.x is None:
+            return 0
```
Comment (tchaton) [inline datasets/base_dataset.py]: Should it be 0 ? If none, it should be sample.pos which is used I think.

File: models/unet_base.py [hunk 2]
```diff
@@ -34,50 +37,48 @@ def __init__(self, opt, model_name, num_classes, modules_lib):
...
                 self._save_sampling_and_search(unet_block, index)
         else:
             index = num_convs
 
         index -= 1
         args_up, args_down = self._fetch_arguments_up_and_down(opt, index, num_convs)
-
-        self.model = UnetSkipConnectionBlock(args_up=args_up, args_down=args_down, output_nc=num_classes, input_nc=None, submodule=unet_block,
-                                             outermost=True, norm_layer=None)  # add the outermost layer
+        args_down['nb_feature'] = dataset.feature_dimension
```
Comment (tchaton) [inline models/unet_base.py]: Great :)

**Comment by tchaton** [comment]
Great feature to the framework. Now, we can benchmark the different models.
Just one minor choice to make for the default input size.

File: datasets/base_dataset.py (hunk 1 above)
Comment (nicolas-chaulet) [reply datasets/base_dataset.py]: Ok, let's do that. If no feature then x,y,z is the feature vector

**Pull request merged**